src/colvaratoms.h

```diff
@@ -492,6 +500,9 @@ class colvarmodule::atom_group
   /// \brief Calculate the derivatives of the fitting transformation
   void calc_fit_gradients();
 
+  template <bool B_ag_center, bool B_ag_rotate>
+  friend void calc_fit_gradients_impl(colvarmodule::atom_group* ag);
```

Comment:
Not exactly. I don't like having a large chunk of template code in a header that's included everywhere, I think the right way to go would be a header that's included only where needed.

Reply:
This function template has to be either a member function, or a friend function of `cvm::atom_group` to access the member variables of  `cvm::atom_group`. If it is the former, then it should be in `colvaratoms.h`, which is included everywhere. If it is the friend function of `cvm::atom_group`, then the definition can be put into `colvaratoms.cpp` directly.

Reply:
The template members should just be _declared_ in `colvaratoms.h`.

Reply:
@jhenin The current code already separates the implementation in `colvaratoms.cpp` (see https://github.com/HanatoK/colvars/blob/0210e5965cc0156d9ceac0314a0c7dbe8e44fea3/src/colvaratoms.cpp#L31-L119). Because it is used only in `colvaratoms.cpp`, I put it there so that it would be instantiated only in the same file (see https://github.com/HanatoK/colvars/blob/0210e5965cc0156d9ceac0314a0c7dbe8e44fea3/src/colvaratoms.cpp#L1315-L1322). I think this function template is unlikely instantiated elsewhere. Do I need to move these lines into a separate header?

The two lines (https://github.com/HanatoK/colvars/blob/0210e5965cc0156d9ceac0314a0c7dbe8e44fea3/src/colvaratoms.h#L503-L504) that you mentioned in `colvaratoms.h` is to make sure that the instantiated function can access the private fields of `cvm::atom_group`.

Reply:
I see. I just pushed a proposed variant that makes it to a class member. It is a very small change, with the only benefit of avoiding the definition of a global function.

Reply:
@jhenin Thanks! I didn't notice that the code would be compiled if the definition is not in the header file, but since the instantiation is only in `colvaratoms.cpp` then defining the template as a member function there should be just fine.